Feature/telegram bot

MaxDubchak opened the pull request:
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/browse/PYTHON-260)


## Code reviewers

- [ ] @meelros
- [ ] @BoroviyOrest
- [ ] @Kardan1123
- [ ] @mizin4ik
- [ ] @dimigor
- [ ] @PetrushynskyiOleksii
- [ ] @Sasha1152

### Second Level Review

- [ ] @lhalam


## Summary of change

added telegram bot handling to backend, and telegram chat access to frontend


## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

MaxDubchak committed:
Add telegram bot handlers, ui chip in user porfile to access telegram bot,
Temporaray saving of bot data to redis.
Added handlers resolve start and help command, chip onClick redirects to telegram link, while payloading user_id to start command of bot.

MaxDubchak committed:
Update telegram bot to save chat id in models, add function to send independent telegram bot message

MaxDubchak committed:
Update bot start handling,
add handling in case of start command without user id attached,
updated bot send message in bot helper

MaxDubchak committed:
Change telegram redirect url to contain token, add token stroing, update telegram id deletion
Add front end token generation of token, token is sent to redis, on response redirects to telegram with token payload
add bothelper functions for redis assistance, bot start command handling changed to token system.

PetrushynskyiOleksii commented on README.md:
```diff
@@ -92,6 +92,10 @@ $ systemctl status rabbitmq-server
     
+    # Required settings for telegram bot
+
+    TELEGRAM_BOT_TOKEN = 'TELEGRAM_BOT_TOKEN'
```
```suggestion
    TELEGRAM_BOT_TOKEN = 'TELEGRAM_BOT_TOKEN'
```
For consistency of readme file, I advise you to change value to `<telegram_bot_token>`.

MaxDubchak replied:
done

PetrushynskyiOleksii commented on way_to_home/telegram_bot/bot_handler.py:
```diff
@@ -0,0 +1,47 @@
+        Function that handles /start command to bot,
+        received data : telegram message, with text property like '/start {user_id}'
+    """
+    if UserProfile.find_by_telegram_id(message.chat.id):
```
I think it will be better to change this method to `get_by_telegram_id`.

MaxDubchak replied:
done

PetrushynskyiOleksii commented on way_to_home/telegram_bot/bot_handler.py:
```diff
@@ -0,0 +1,47 @@
+        BOT.send_message(chat_id=message.chat.id,
+                         text=bothelper.UNAUTHORIZED_ACCESS_MESSAGE)
+    else:
+        user = CustomUser.get_by_id(bothelper.get_user_id_by_access_token(token))
```
I think it will be better to take off `get_user_id_by_access_token`:
```python
user_id = bothelper.get_user_id_by_access_token(token)
user = CustomUser.get_by_id(user_id)
```

MaxDubchak replied:
done

PetrushynskyiOleksii commented on way_to_home/telegram_bot/bot_handler.py:
```diff
@@ -0,0 +1,47 @@
+            return
+
+        user.user_profile.update(telegram_id=message.chat.id)
+        bothelper.remove_redis_access_token(user)
```
Maybe will be better `remove_user_access_token`?

PetrushynskyiOleksii commented on way_to_home/telegram_bot/bothelper.py:
```diff
@@ -0,0 +1,68 @@
+    """This function sends telegram message with given text to user"""
+    chat_id = user.user_profile.telegram_id
+    if not chat_id:
+        return False
```
```suggestion
        def send_telegram_message(text, user):
        """This function sends telegram message with given text to user"""
            chat_id = user.user_profile.telegram_id
            if not chat_id:
                return False
```
In our case, you need here just logic of sending message.
So change param `user` to `chat_id`. Logic of retrieve `chat_id` must be in celery task.

MaxDubchak replied:
right, done

PetrushynskyiOleksii commented on way_to_home/telegram_bot/bothelper.py:
```diff
@@ -0,0 +1,68 @@
+    return True
+
+
+def get_redis_access_tokens():
```
It is better to change to `get_access_tokens`.
I think the description of how and where it retrieves must leave in docstrings.
Exceptionally, you don't store `access_tokens` in one place.

PetrushynskyiOleksii commented on way_to_home/telegram_bot/bothelper.py:
```diff
@@ -0,0 +1,68 @@
+def get_user_id_by_access_token(token_check):
+    """
+        find user_id by access token in redis
+    """
```
Can you please write it in single line?

MaxDubchak replied:
done, that was an accident

PetrushynskyiOleksii commented on way_to_home/user_profile/models.py:
```diff
@@ -42,3 +42,27 @@ def create(cls, user, first_name='', last_name='', telegram_id=None):  # pylint:
+                return False
+
+    @classmethod
+    def find_by_telegram_id(cls, telegram_id):
```
As I wrote above, please change it to `get_by_telegram_id`

MaxDubchak replied:
as i wrote above, done)

PetrushynskyiOleksii commented on way_to_home/user_profile/urls.py:
```diff
@@ -2,9 +2,11 @@
 
 urlpatterns = [
     path('', UserProfileView.as_view(), name='profile'),
+    path('/telegram_redis', put_key_to_redis, name='telegram_redis'),
```
Maybe it will be better to change url path to `/telgram_access_token`

PetrushynskyiOleksii commented on way_to_home/user_profile/urls.py:
```diff
@@ -2,9 +2,11 @@
 
 urlpatterns = [
     path('', UserProfileView.as_view(), name='profile'),
+    path('/telegram_redis', put_access_token, name='telegram_redis'),
```
It is better to change it to `telegram_access_token`.
Because URL it is a public thing and it is not good to show where we store `access token` xD

MaxDubchak replied:
done

MaxDubchak committed:
Update celery task 'send_notification'
changed send notificiation task to try and send notifciation to telegram first, then to phone number, aand to retry up to 5 times on failure

MaxDubchak committed:
removed google token field from user model
